Source/WebCore/rendering/RenderLayer.cpp

```diff
@@ -1077,14 +1080,18 @@ void RenderLayer::updateLayerPositionsAfterLayout(RenderElement::LayoutIdentifie
     gVisitedPositionsCount = 0;
 #endif
 
-    LOG(Compositing, "RenderLayer %p updateLayerPositionsAfterLayout - before", this);
+    LOG_WITH_STREAM(Compositing, stream << "RenderLayer " << this << " updateLayerPositionsAfterLayout (environment changed " << environmentChanged << " - before");
 #if ENABLE(TREE_DEBUGGING)
     if (compositingLogEnabledRenderLayer())
         showLayerPositionTree(root());
 #endif
+
+    if (environmentChanged)
+        setSelfAndDescendantsNeedPositionUpdate();
```

Comment:
We shouldn't need to do both of setting the flag and marking the layer. Either are sufficient to trigger recomputation of layer positions of the subtree.

The general idea is that the 'need position update' bits (self, self+children, self+alldescendants) are for changes that happen outside of updating layer positions. The flags are for state changes that are detected during the walk, and we newly need to walk descendants.

The flags could probably just be replaced by adding new dirty bits mid-walk, but they have existing meaning too.